Helpers

sethjeffery opened the pull request:
From the README...


## Helper methods

You can define helpers with the `helper` keyword. Helpers can be used within blocks, procs, or as symbolic names for if statements to tidy up your code.

```rb
Flagship.define :blog do
  helper :is_author do |comment, user|
    comment.author == user
  end

  def can_view_comment(context)
    context.current_user.moderator?
  end

  enable :comment, if: :can_view_comment
  enable :comment_deletion, if: ->(context) { is_author(context.comment, context.current_user) }
end
```

To share helpers, you can simply include them as modules.


```rb
module FlagHelpers
  def is_author(context)
    context.comment.author == context.current_user
  end
end

Flagship.define :development do
  include FlagHelpers
  enable :delete, if: :is_author
end

Flagship.define :production do
  include FlagHelpers
  enable :delete, if: :is_author
end
```

sethjeffery commented:
Yes one main difference is that `helper` methods are kept within the context of the flagset, but if you define by `def` it is stored in the Flagship global context shared with all flagsets, which I find a bit dangerous because you can accidentally overwrite definitions.

yuya-takeyama commented:
@sethjeffery 
>if you define by `def` it is stored in the Flagship global context shared with all flagsets

Oh, that's dangerous...
But I couldn't see such a situation.

```rb
require 'flagship'

Flagship.define :a do
  def foo(c)
    puts "foo is called"
    true
  end
end

Flagship.define :b do
  enable :feature, if: ->(c) { foo(c) }
end

Flagship.select_flagset(:b)

p Flagship.enabled?(:feature)
```

```
$ bundle exec ruby test.rb
test.rb:11:in `block (2 levels) in <main>': undefined method `foo' for #<Flagship::Dsl:0x007f96e3223090> (NoMethodError)
Did you mean?  fork
        from /Users/yuya/src/github.com/sethjeffery/flagship/lib/flagship/feature.rb:24:in `enabled?'
        from /Users/yuya/src/github.com/sethjeffery/flagship/lib/flagship/flagset.rb:17:in `enabled?'
        from /Users/yuya/src/github.com/sethjeffery/flagship/lib/flagship.rb:20:in `enabled?'
        from test.rb:16:in `<main>'
```

sethjeffery commented:
Hmm you're right! I actually had originally written to use `def` but I encountered problems so I changed it... now I'm not sure how to reproduce the problems! (Maybe I fixed it through another change?)

yuya-takeyama commented:
Maybe by this change?
https://github.com/yuya-takeyama/flagship/pull/21/commits/2817e99befe86380c24cc8d7883628a9c7ab6cea

Then I think we don't need `helper` keyword.
Basically I want to keep the library small.

sethjeffery commented:
Ok I'll remove the keyword.

yuya-takeyama commented:
I'll release this with one more change later today.
